### PR TITLE
Refactor `deduplicate()` arguments

### PR DESCRIPTION
**Pull request by judahrand**
This is a:
- [ ] bug fix PR with no breaking changes — please ensure the base branch is `main`
- [ ] new functionality — please ensure the base branch is the latest `dev/` branch
- [ ] a breaking change — please ensure the base branch is the latest `dev/` branch

## Description & motivation

This Pull Request aims to improve the `deduplicate` macro based on community feedback from its first release. It is stacked on top of #549 and introduces breaking changes.

- #542 discusses argument naming as well as how to improve the documentation.
- #542 also raises the issue that the usage of `relation_alias` to reference CTEs is confusing. This prompted me to investigate whether this is really needed - it was only needed due to the use of `dbt_utils.star` which can be avoided by using a `natural join` which is supported by most SQL syntaxes.

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [x] BigQuery
    - [x] Postgres
    - [x] Redshift
    - [x] Snowflake
- [x] I followed guidelines to ensure that my changes will work on "non-core" adapters by:
    - [x] dispatching any new macro(s) so non-core adapters can also use them (e.g. [the `star()` source](https://github.com/fishtown-analytics/dbt-utils/blob/master/macros/sql/star.sql))
    - [x] using the `limit_zero()` macro in place of the literal string: `limit 0`
    - [x] using `dbt_utils.type_*` macros instead of explicit datatypes (e.g. `dbt_utils.type_timestamp()` instead of `TIMESTAMP`
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have added an entry to CHANGELOG.md

### CODE REVIEW

**Comment by judahrand** [comment]
I'm not sure why the Redshift test is failing with
```
09:05:22  Database Error in model test_deduplicate (models/sql/test_deduplicate.sql)
09:05:22    SELECT DISTINCT ON is not supported
09:05:22    compiled SQL at target/run/dbt_utils_integration_tests/models/sql/test_deduplicate.sql
09:05:22  Encountered an error:
FailFast Error in model test_deduplicate (models/sql/test_deduplicate.sql)
  Failing early due to test failure or runtime error
```
given that `DISTINCT ON` only appears in the Postgres implementation...

**Comment by judahrand** [comment]
Ah, didn't realize Redshift defaults to Postgres rather than default.

File: README.md
Comment (dbeatty10) [inline README.md:750]:
```suggestion
    partition_by="user_id, cast(timestamp as day)",
```

File: README.md
Comment (dbeatty10) [inline README.md:765]:
```suggestion
    partition_by="user_id, cast(timestamp as day)",
```

File: README.md
Comment (dbeatty10) [inline README.md:755]: Love that there are usage examples for:
* a `source()` relation
* CTE
* `partition_by` containing multiple expressions

To round out the usage examples, it would be awesome to include:
* a `ref()` relation
* `order_by` containing multiple expressions

Shooting from the hip, the following attempts to do both in one example:
```sql
{{ dbt_utils.deduplicate(
    relation=ref('my_model'),
    partition_by="user_id",
    order_by="effective_date desc, effective_sequence desc",
   )
}}
```

**Review by dbeatty10 — COMMENTED** [review]
@judahrand love the direction this is heading! Thank you for your detailed coverage of a wide range of databases, especially as it relates to using natural joins.

The updates are definitely more readable.

I left a few comments regarding the documentation -- appeared that the renaming to `partition_by` was overlooked in the usage examples.

I'm wondering if we can do any fancy footwork in the background so that the changes aren't breaking (but still make all the updates we want). I think I've seen some examples in past merge requests, so I'm going to try looking for those, and I'll follow-up in this conversation if/when I find them.

**Comment by judahrand** [comment]
@dbeatty10 I'll deal with these comments now 😄 I also wasn't sure if you thought this would have to wait for a minor release due to breaking changes even though the macro is very very new??

File: README.md
Comment (codigo-ergo-sum) [inline README.md:743]: Might I suggest an additional sentence here? "If there are multiple rows within the window specified by `partition_by` which share the same values for the columns specified in `order_by` then the result of deduplication may be nondeterministic over multiple runs of the model using the macro given that there is no way for the database to pick which row should 'win' the tie."

Here's an example of what I am talking about, it's for Snowflake, but it will be mathmatically equivalent for all other database platforms because it's not a specific database implementation thing, it's a math thing: https://community.snowflake.com/s/article/ROW-NUMBER-function-causing-Non-deterministic-behavior

Comment (codigo-ergo-sum) [reply README.md:743]: And I'm actually seeing this exact behavior in one of the current dbt projects that I am working on where someone used `ROW_NUMBER` to deduplicate but neglected to check if there could be "ties". So it's not just a rare or theoretical problem.

Comment (judahrand) [reply README.md]: Added a small note. Happy to call this resolved?

Comment (codigo-ergo-sum) [reply README.md]: This is great thank you so much :).

**Comment by judahrand** [comment]
@dbeatty10 I'd like to try to move ahead with this? Did you manage to find any examples of the fancy footwork you mentioned?

**Comment by judahrand** [comment]
@dbeatty10 I'd really like to move forward with this if possible?

**Comment by dbeatty10** [comment]
@judahrand thank you for your work on this -- I can dedicate time this week to help overcome any barriers.

One big thing has changed over the last couple weeks -- we're targeting all contributions for the `main` branch now. This is a change from using the  `next/*` branches. Hopefully this makes things easier for folks down the road. But will probably make things a little trickier for you in the meantime. Just let me know if there's anything I can do to help with that piece. 

I think there's two main things we'll need to do to merge this in once all tests are passing:
1. Make sure there's no breaking changes (if possible)
2. Be able to merge to `main` instead of `next/patch`

**Comment by judahrand** [comment]
> I think there's two main things we'll need to do to merge this in once all tests are passing:
> 
> 1. Make sure there's no breaking changes (if possible)
> 2. Be able to merge to `main` instead of `next/patch`

Hey @dbeatty10,

I think that I've managed to overcome both of these barriers thanks to your pointers over Slack.

Are you able to re-review this? 

Once this is merged it would also be great to try to get #550 and #551 in too.

File: CHANGELOG.md
Comment (dbeatty10) [inline CHANGELOG.md:8]:
```suggestion
- The `order_by` argument is now required.
```

File: CHANGELOG.md
Comment (dbeatty10) [inline CHANGELOG.md:38]:
```suggestion
- Add better documentation for `deduplicate()` ([#542](https://github.com/dbt-labs/dbt-utils/pull/542), [#548](https://github.com/dbt-labs/dbt-utils/pull/548))
```

**Review by dbeatty10 — CHANGES_REQUESTED** [review]
This looks great @judahrand ! Thanks for your documentation of how to handle the deprecated signature and for adding an integration test for that case.

I left two comments on very minor things to update within the documentation.

Looks good to me to merge after that.

**Comment by judahrand** [comment]
@dbeatty10 Sorted!

**Pull request merged**
